First commit with distance

glee2244 opened the pull request:
I've attempted to include distance in the styling language for vectors. But I'm still confused as to how Cartesian3.distance() could work when the inputs could be anything other than type vec3. I've also tried to run my tests on my local server but the SpecRunner doesn't seem to display anything.
Issue #4865 

@Dylan-Brown

lilleyse commented:
@glee2244 in the future make a branch off of 3d-tiles and do your work there, this will make it easier to sync up your 3d-tiles branch with the origin.

lilleyse commented:
> I've attempted to include distance in the styling language for vectors. But I'm still confused as to how Cartesian3.distance() could work when the inputs could be anything other than type vec3.

Ok yeah I see the problem now, `distance`, `dot`, and `cross` will need to be handled differently than the other binary functions because they are geometric operations / don't work component-wise like the others.

Probably the best thing to do is be consistent with what's already there and set distance to point to a local function. (Example: https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/Source/Scene/Expression.js#L107-L116). The function will just compute the distance between two numbers.

Then inside `getEvaluateBinaryFunction` there will need to be a special case for `distance`. It will be a separate block that looks similar to the if/else already there. For the number/number check, use the evaluate function that you set above. For the vector cases, use `Cartesian2/3/4.distance` on the `left` and `right`.

lilleyse commented:
> I've also tried to run my tests on my local server but the SpecRunner doesn't seem to display anything.

You may just need to build, try running `npm run build`. This is important whenever switching between branches that have different files, like going between master and 3d-tiles.

lilleyse commented:
@glee2244 Do you have enough to go with here?

glee2244 commented:
@lilleyse Yes, thanks so much for the pointers!

glee2244 commented:
@lilleyse Should I be throwing a Developer Error if the distance function is called with mismatching inputs? For example, calling the distance between a number and a vec3?

lilleyse commented:
Yes, throw an error for those cases.

glee2244 commented:
@lilleyse I've pushed the changes as you've suggested, and it seems like the tests I've written pass, but I still get a failed check - Travis CI Build failed? Any ideas here?

lilleyse commented:
There are some unrelated test failures in the `3d-tiles` branch right now, just ignore for now.

glee2244 commented:
@lilleyse Alright, duly noted. I'm going to continue with implementing dot and cross.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -1,28 +1,28 @@
+    isArray,
+    CesiumMath,
+    jsep,
+    ExpressionNodeType) {
```
Undo this indentation changes in here and `ExpressionSpec.js`.

However the rest of the formatting fixes seem fine.

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -2143,6 +2143,30 @@ defineSuite([
         }).toThrowDeveloperError();
     });
 
+    it ('evaluates the distance function', function() {
```
`it (` -> `it(` here and below.

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -2143,6 +2143,30 @@ defineSuite([
+            return new Expression('distance(1, 3, 0)');
+        }).toThrowDeveloperError();
+    });
+
```
Also adds tests for mismatching types.

glee2244 commented:
Pushing my implementations to dot and cross, but some of my tests are failing - will resolve them within the next couple days.

glee2244 commented:
@lilleyse Is cross meant to only take two parameters? I was trying to call Cartesian3.cross with two parameters but just realized that the function I wanted to use takes 3 (the third being the object onto which to store the result). https://cesiumjs.org/Cesium/Build/Documentation/Cartesian3.html?classFilter=Carte

lilleyse commented:
For that you can send in `ScratchStorage.getCartesian3()`

glee2244 commented:
@lilleyse HI Sean, I've fixed my tests and some indentation/spacing that you pointed out before, and I've also refactored my implementations slightly. Let me know if there's anything else I should fix.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -123,6 +126,44 @@ define([
 
     function log2(number) {
         return CesiumMath.logBase(number, 2.0);
+
```
Remove empty line.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -123,6 +126,44 @@ define([
+        } else {
+            throw new DeveloperError('Function dot requires matching types of inputs. Arguments are ' + x + ' and ' + y + '.');
+        }
+
```
Remove empty line.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -123,6 +126,44 @@ define([
+        if (typeof x === 'number' && typeof y === 'number') {
+            return Math.abs(x - y);
+        } else if (x instanceof Cartesian2 && y instanceof Cartesian2) {
+            return Cartesian2.distance(x ,y);
```
Formatting fix: `(x ,y)` -> `(x, y)`

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -123,6 +126,44 @@ define([
+        } else if (x instanceof Cartesian4 && y instanceof Cartesian4) {
+            return Cartesian4.distance(x, y);
+        } else {
+            throw new DeveloperError('Function distance requires matching types of inputs. Arguments are ' + x + ' and ' + y + '.');
```
Wrap each of these three `DeveloperError` with 

```
//>>includeStart('debug', pragmas.debug);
throw new DeveloperError('.....')
//>>includeEnd('debug');
```

so that it is not included in release builds.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -1737,11 +1784,11 @@ define([
                 //>>includeStart('debug', pragmas.debug);
                 throw new DeveloperError('Error generating style shader: undefined is not supported.');
-                //>>includeEnd('debug');
+            //>>includeEnd('debug');
```
Remove the indentation changes here.

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -1,20 +1,20 @@
+    Cartesian4,
+    Color,
+    CesiumMath,
+    ExpressionNodeType) {
```
Reminder to revert the indentation for the includes here and `Expression.js`

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -2143,6 +2143,105 @@ defineSuite([
+    it('throws if dot function takes an invalid number of arguments', function() {
+        expect(function() {
+            return new Expression('dot(0.0)');
+        }) .toThrowDeveloperError();
```
Fix formatting for `}) .toThrowDeveloperError();`

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -2143,6 +2143,105 @@ defineSuite([
+
+    it('throws if distance function takes mismatching types of arguments', function() {
+        expect(function() {
+            return new Expression('distance(1, vec2(3.0, 2.0)').evaluate(frameState,undefined);
```
Style: `frameState,undefined` to `frameState, undefined` for this and others.

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -2143,6 +2143,105 @@ defineSuite([
+        }).toThrowDeveloperError();
+    });
+
+    it('throws if cross function takes mismatching types of arguments', function() {
```
Rename to `throws if cross function does not take vec3 arguments`

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -2167,6 +2266,7 @@ defineSuite([
         }).toThrowDeveloperError();
     });
 
+
```
Remove empty line.

lilleyse commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -3016,7 +3116,7 @@ defineSuite([
+    it('gets shader expression for sin', function() {
         var expression = new Expression('sin(0.0)');
         var shaderExpression = expression.getShaderExpression('', {});
         var expected = 'sin(0.0)';
```
Can you also add tests for `distance`, `dot`, and `cross` for the shader expressions? Similar to 

```
it('gets shader expression for min', function() {
    var expression = new Expression('min(3.0,5.0)');
    var shaderExpression = expression.getShaderExpression('', {});
    var expected = 'min(3.0, 5.0)';
    expect(shaderExpression).toEqual(expected);
});
```

glee2244 commented:
Have added shader expression tests for distance, dot and cross and fixed spacing/indentation as requested, except for reverting indentation for includes here - for some odd reason the indentation keeps going back to incorrect spacing even after I update and save it. Will try to resolve asap.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -125,6 +128,46 @@ define([
+        if (x instanceof Cartesian3 && y instanceof Cartesian3) {
+            return Cartesian3.cross(x, y, ScratchStorage.getCartesian3());
+        } else {
+            throw new DeveloperError('Invalid argument type for cross() function, must be vec3');
```
Wrap this with `//>>includeStart('debug', pragmas.debug);` and `//>>includeEnd('debug');` too

lilleyse replied:
This should also be wrapped with the debug pragma.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -1623,7 +1672,7 @@ define([
                 //>>includeStart('debug', pragmas.debug);
                 throw new DeveloperError('Error generating style shader: "' + value + '" is not supported.');
-                //>>includeEnd('debug');
+            //>>includeEnd('debug');
```
Undo all these indentations as well.

lilleyse commented on Source/Scene/Expression.js:
```diff
@@ -125,6 +128,48 @@ define([
+        if (typeof x === 'number' && typeof y === 'number') {
+            return x * y;
+        } else if (x instanceof Cartesian2 && y instanceof Cartesian2) {
+            return Cartesian2.dot(x ,y);
```
Change to `(x, y)`

lilleyse commented on Source/Scene/PointCloud3DTileContent.js:
```diff
@@ -872,6 +875,7 @@ define([
         var hasNormals = content._hasNormals;
         var hasBatchIds = content._hasBatchIds;
         var backFaceCulling = content._backFaceCulling;
+        //var normalShading = content._normalShading;
```
Remove this commented `normalShading` code.

lilleyse commented:
Can you remove the `normalShading` code entirely? That should only belong in the other PR.

lilleyse commented:
Thanks @glee2244!
